lib/apipie/swagger_generator.rb

```diff
@@ -0,0 +1,583 @@
+    #--------------------------------------------------------------------------
+
+    def ruby_name_for_method(method)
+      return "<no method>" if method.nil?
+      method.resource.controller.name + "#" + method.method
+    end
+
+
+    def warn_missing_method_summary() warn 100, "missing short description for method"; end
+    def warn_added_missing_slash(path) warn 101,"added missing / at beginning of path: #{path}"; end
+    def warn_no_return_codes_specified()  warn 102,"no return codes ('errors') specified"; end
+    def warn_hash_without_internal_typespec(param_name)  warn 103,"the parameter :#{param_name} is a generic Hash without an internal type specification"; end
+    def warn_optional_param_in_path(param_name) warn 104, "the parameter :#{param_name} is 'in-path'.  Ignoring 'not required' in DSL"; end
+    def warn_optional_without_default_value(param_name) warn 105,"the parameter :#{param_name} is optional but default value is not specified (use :default_value => ...)"; end
+    def warn_param_ignored_in_form_data(param_name) warn 106,"ignoring param :#{param_name} -- cannot include Hash without fields in a formData specification"; end
+    def warn_path_parameter_not_described(name,path) warn 107,"the parameter :#{name} appears in the path #{path} but is not described"; end
+
+
+    def warn(warning_num, msg)
+      suppress = Apipie.configuration.swagger_suppress_warnings
+      return if suppress == true
+      return if suppress.is_a?(Array) && suppress.include?(warning_num)
+
+      method_id = ruby_name_for_method(@current_method)
+      warning_id = "#{method_id}#{warning_num}#{msg}"
+
+      if @issued_warnings.include?(warning_id)
+        # print "--> Already issued warning: [#{warning_id}]\n"
+        return
+      end
+
+      print "WARNING (#{warning_num}): [#{method_id}] -- #{msg}\n"
+      @issued_warnings.push(warning_id)
+      @warnings_issued = true
+    end
+
+    def info(msg)
+      print "--- INFO: [#{ruby_name_for_method(@current_method)}] -- #{msg}\n"
+    end
+
+    def investigate(msg)
+      # print "??? RESEARCH: [#{ruby_name_for_method(@current_method)}] -- #{msg}\n"
+    end
+
+    def dump_param_desc(param_desc)
+      methods = param_desc.methods - param_desc.class.methods
+      for m in methods
+        print "#{m}\n"
+      end
+    end
+
+
+    # the @master_id is a number that is uniquely derived from the list of operations
+    # added to the swagger definition (in an order-dependent way).
+    # it is used for regression testing, allowing some differentiation between changes that
+    # result from changes to the input and those that result from changes to the generation
+    # algorithms.
+    # note that at the moment, this only takes operation ids into account, and ignores parameter
+    # definitions, so it's only partially useful.
+    def include_op_id_in_master_identifier(op_id)
+      @master_id = Zlib::crc32("#{@master_id} #{op_id}")
```

Comment:
we should probably require `zlib` somewhere: it's probably loaded by rails somewhere, but just to be on safe side